### DISCUSSION TITLE
Bug: Not respecting CORS policy causes blocked requests

### DISCUSSION THREAD

**Discussion by account00001** [SixLabors/ImageSharp.Web#291] [Q&A] [answered]
I can access my image via URL fine but when I add processing commands to it, for example `?width=400`, the request is blocked by CORS. I have my processing behavior set to `ProcessingBehavior.CommandOnly`.

Example of successful request without processing:
![image](https://user-images.githubusercontent.com/52767641/196062861-12cb9d16-e84e-4615-b149-da6fd5158e73.png)

Example of CORS blocking with processing:
![image](https://user-images.githubusercontent.com/52767641/196062872-30d2b1a8-b9c8-4bf1-9c38-8df9c7302865.png)

Example when I change processing behavior to `ProcessingBehavior.All` (request is blocked without processing)
![image](https://user-images.githubusercontent.com/52767641/196063094-c31ce7d5-a899-4b24-b07b-c36dac56ddaa.png)

It's blocked because the CORS header ‘Access-Control-Allow-Origin’ is missing on the response. I've configured my project to be setup with CORS via `builder.Services.AddCors(...)`. I've added a default policy and allowed my origins. It seems responses generated by ImageSharp however are not respecting this policy.

I've searched the documentation and example projects but I couldn't find anything related to CORS.

Questions:

How can I configure CORS when I call `services.AddImageSharp(...)`?
or
How can I get ImageSharp to respect my existing CORS policy?

I'm on SixLabors.ImageSharp.Web v2.0.2 with .NET6. Happens in all browsers (tested with Firefox, Chrome, Edge).

**Comment by JimBobSquarePants** [comment, marked as answer]
You can use [`ImageSharpMiddlewareOptions.OnPrepareResponseAsync`](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddlewareOptions.html#SixLabors_ImageSharp_Web_Middleware_ImageSharpMiddlewareOptions_OnPrepareResponseAsync) to further augment the response headers before sending.

It would be good to automate that though.

**Reply by account00001** [reply]
Thanks for that work around. I got it working by mimicking my CORS policy in `OnPrepareResponseAsync` by setting `AccessControlAllowOrigin`, `AccessControlAllowHeaders`, and `AccessControlAllowMethods` headers on my response.

However, like you said, it would be nice if there was more of a built-in/fluent support for configuring CORS.

Thanks

**Reply by JimBobSquarePants** [reply]
Can you raise an issue in the tracker please or else I'll forget. Thanks!

**Reply by JimBobSquarePants** [reply]
Actually. Can I see your startup code? I want to see the order of registration of middleware

**Reply by account00001** [reply]
Trimmed for brevity.

```csharp
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddCors();

builder.Services.AddControllers();

builder.Services.AddImageSharp();

var app = builder.Build();

app.UseDefaultFiles();
app.UseImageSharp();
app.UseStaticFiles();

app.UseCors();

app.UseRouting();
app.MapControllers();

app.Run();
 ```

**Reply by JimBobSquarePants** [reply]
Where's `app.UseImageSharp();`?

**Reply by account00001** [reply]
Sorry, I had it as `UseImageApi`, `app.UseImageSharp` is between Default files and Static files. Moving `app.UseCors()` above `app.UseImageSharp()` might fix the CORS issues, but I can't say for certain because I get exceptions when retrieving the images.

**Reply by JimBobSquarePants** [reply]
I think it’s supposed to be last. I’m gonna have a dig through the ASP.NET source

**Reply by JimBobSquarePants** [reply]
Just to check again... `app.UseImageSharp(app.Environment);` isn't correct. There's no overload of that method that takes the application environment.

**Reply by JimBobSquarePants** [reply]
I'm not seeing this while testing it using the sample project in the repository.

```c#
/// <summary>
/// The running application.
/// </summary>
public static class Program
{
    /// <summary>
    /// The main application entry point.
    /// </summary>
    /// <param name="args">Argument paramateres.</param>
    public static void Main(string[] args)
    {
        WebApplicationBuilder builder = WebApplication.CreateBuilder(args);

        // Add services to the container.
        IServiceCollection services = builder.Services;
        services.AddRazorPages();

        // TODO: Enable HMAC
        services.AddImageSharp(options => options.HMACSecretKey = null) // new byte[] { 1, 2, 3, 4, 5 })
            .SetRequestParser<QueryCollectionRequestParser>()
            .Configure<PhysicalFileSystemCacheOptions>(options =>
            {
                options.CacheRootPath = null;
                options.CacheFolder = "is-cache";
                options.CacheFolderDepth = 8;
            })
            .SetCache<PhysicalFileSystemCache>()
            .SetCacheKey<UriRelativeLowerInvariantCacheKey>()
            .SetCacheHash<SHA256CacheHash>()
            .Configure<PhysicalFileSystemProviderOptions>(options => options.ProviderRootPath = null)
            .AddProvider<PhysicalFileSystemProvider>()
            .AddProcessor<ResizeWebProcessor>()
            .AddProcessor<FormatWebProcessor>()
            .AddProcessor<BackgroundColorWebProcessor>()
            .AddProcessor<QualityWebProcessor>()
            .AddProcessor<AutoOrientWebProcessor>();

        services.AddCors();

        WebApplication app = builder.Build();

        // Configure the HTTP request pipeline.
        if (!app.Environment.IsDevelopment())
        {
            app.UseExceptionHandler("/Error");

            // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
            app.UseHsts();
        }

        app.UseHttpsRedirection();

        app.UseImageSharp();

        app.UseStaticFiles();

        app.UseCors();

        app.UseRouting();

        app.UseAuthorization();

        app.MapRazorPages();

        app.Run();
    }
}
```
<img alt="image" src="https://user-images.githubusercontent.com/385879/196422596-44fc0082-9575-4f1a-ba2e-e686c8de4914.png">

**Reply by account00001** [reply]
I wrapped ImageSharp up in my own wrapper, the `app.Environment` is residual from that and has been removed. Sorry for the confusion.

As far as the reproduction example, I don't think that would reproduce it because you are on the same origin. The issue arises when you attempt a cross origin request (UI is on port A and API is on port B).

**Reply by account00001** [reply]
# Setup
Add an image called "image.png" to your desktop.

In the "frontend" directory, run `npm install` then `npm run serve`. This assumes you have NPM And NodeJS installed.

In the "backend" directory, simply hit the play button in Visual Studio with "Api" as your startup project.

# Frontend
Visit https://localhost:8080/ to load "image.png" from the frontend (UI, port 8080).
![image](https://user-images.githubusercontent.com/52767641/196599717-0e8c8e80-3fbf-42d9-a2d1-95941f79b3e2.png)
![image](https://user-images.githubusercontent.com/52767641/196599915-8ff92e9b-cc3d-4a0d-93fb-b05ebf60d37d.png)
Notice we are on port 8080 making a request to 7227. Cross origin.

# Backend
Visit https://localhost:7227/api/image to load "image.png" from the backend (API, port 7227).
![image](https://user-images.githubusercontent.com/52767641/196599768-f71a870a-53c0-457e-91d7-023f664682e2.png)
![image](https://user-images.githubusercontent.com/52767641/196600011-ac185bd5-a126-4905-8619-0aab82bf5fed.png)
We are on 7227 making a request to 7227. Ok.

# Reproduction

With lines 31-37 commented out in `Program.cs`, relaunch the Api project and the CORS error occurs (request is blocked).
```csharp
            builder.Services.AddImageSharp(options =>
            {
                //options.OnPrepareResponseAsync = (context) =>
                //{
                //    context.Response.Headers.AccessControlAllowOrigin = new("https://localhost:8080");
                //    context.Response.Headers.AccessControlAllowHeaders = new(new[] { "content-type", "cache-control" });
                //    context.Response.Headers.AccessControlAllowMethods = new("GET");
                //    return Task.FromResult(context.Response);
                //};
            })
```
![image](https://user-images.githubusercontent.com/52767641/196599428-d5d36e08-808d-4dd9-94c5-8dee7ccf5d9c.png)

# Files
Here is the complete example.

[ImageSharp.Web-discussions-291.zip](https://github.com/SixLabors/ImageSharp.Web/files/9816906/ImageSharp.Web-discussions-291.zip)

**Reply by JimBobSquarePants** [reply]
Thanks for this. It made it nice and easy to replicate the issue!

I have a fix in the works which will apply the correct policy automatically.

**Reply by account00001** [reply]
Adding to this, I think there may be a more complex issue afoot. I've been seeing odd caching behavior, even after disabling all the cache and using a `NullImageCache` with `SetCache`. Completely disabling ImageSharp fixes the caching issue. I've managed to trace it back to relating to CORS.

The CORS preflight request sends an OPTIONS request to the backend and it should respond with a 204 No Content. Then it sends the GET request to the same endpoint to retrieve the content.

What I'm seeing is the preflight OPTIONS request return a 200 with some content. Later, the GET request also returns a 200 with some content. 

These two requests are hitting the same endpoint with the same image ID and should be returning the same image, however both the OPTIONS and the GET are returning a random image (the image isn't even consistent between the same request, meaning both the preflight and the actual request can be different images while attempting to retrieve the same ID, plus preflight shouldn't return anything to begin with).

When I access the image directly from the backend it is always correct. 

I'm not sure why the OPTIONS is returning actual content, but I think it's throwing something off when the actual GET request is sent. I think part of respecting the CORS policy should also include appropriate handling for OPTIONS requests, as right now it seems it's not looking at the request method before returning a response. As far as it returning a random image even with all cache disabled, I have no idea. It probably returns a random image like 30% of the time.

I'm not really sure where to go from here, I suppose I'll have to disable ImageSharp until I can test the CORS fix and see if it also fixes this issue. Please let me know when the fix is ready, I'd be interested in helping test it.

**Comment by JimBobSquarePants** [comment]
These symptoms are unlike anything I have ever heard and the middleware is used on thousands of sites serving millions of images. With the default hashing strategy in place collision opportunities are 1 in 10s of millions. 

I haven’t forgotten about the CORS fix. I have local code that returns the correct response I’m just not sure whether the middleware is the best place to do so. Normally CORS behaviours are configured against individual endpoints in the routing configuration.

**Reply by account00001** [reply]
Good news, I managed to track down the root issue being bad async code inside the image provider. Now that that's fixed, I no longer get random images. This was positively interfering with the caching behavior on top of the preflight issue, causing some red herrings. Bad news is the CORS issue where the preflight returns a 200 with a response is still a true bug, however much less severe (unless resources/bandwidth are important). I expect that should be an easy fix, to just check the method on the request before going through the rest of the pipeline. I can give you an example project which reproduces that if needed.
